### PR TITLE
Adding box plots drawn on std-deviation instead of quartiles

### PR DESCRIPTION
**Pull request by 28raining**
To help with this issue
https://github.com/plotly/plotly.js/issues/6697

I've made below changes, I've tried to keep the change to the minimum. Resulting image of new plot is in the issue

- To box plot I've added a new boolean setting "whiskerdisable", which will turn of whiskers and whisker caps. Unfortunately whiskerwidth = 0 (existing setting) only turns off whisker caps
- I added 6 new options to box -> boxmean setting, for 1 thru 6 sigma plots
- On hover the annotation shows mean +- sigma for these new settings
- Added new mock called box_3sigma

I also touched CONTRIBUTING.md because this instructions opens an empty webpage and it's really not obvious how to use it

### CODE REVIEW

File: src/traces/box/attributes.js [hunk 1]
```diff
@@ -221,12 +221,14 @@ module.exports = {
 
     boxmean: {
         valType: 'enumerated',
-        values: [true, 'sd', false],
+        values: [true, 'sd', '1sigma', '2sigma', '3sigma', '4sigma', '5sigma', '6sigma', false],
```
Comment (alexcjohnson) [inline src/traces/box/attributes.js:224]: My gut reaction is `boxmean` - which currently is just about adding a little extra information - isn't the right place to describe such a foundational feature as "what does the size of the box mean," especially since `true` is already taken.

I'd propose we make two new attributes:
- `sizemode` - defaults to `quartiles` (current behavior) but you can set `sd` to change to mean ± N sd
- `sdmultiple` - a number rather than an enumeration of a few values. Only available if `sizemode='sd'` or `boxmean='sd'`, defaults to `1` and has minimum `0`, and multiplies sd in either case (box or diamond).

And then if you set `sizemode='sd'`, the `boxmean` attribute is not available.

File: src/traces/box/attributes.js [hunk 2]
```diff
@@ -378,6 +380,15 @@ module.exports = {
         ].join(' ')
     },
 
+    whiskerdisable: {
+        valType: 'boolean',
+        dflt: false,
```
Comment (alexcjohnson) [inline src/traces/box/attributes.js:385]: Let's avoid the double negative: call the attribute `showwhiskers`. And assuming that there isn't normally a purpose for whiskers in this new "mean ±N sd" mode, we can make the default `true` when we're showing quartiles and `false` for sd. To do this, you calculate the default in `defaults.js` after coercing the attribute(s) it depends on, like we do with `boxmeanDflt`, and describe this logic in words in the below `description`.

What happens if you DO show whiskers with mean/sd boxes? Are they still at the upper/lower fence positions even though those don't mean much without the quartiles?

**Comment by 28raining** [comment]
Hi @alexcjohnson thank you for the feedback.
- whiskerdisable renamed to showwhiskers. Mock test now includes with and without whiskers. Sometimes the box is larger than the data so the whisker draws inwards. Changing that behavior might have unintended side effects, and I don't think it looks too bad
- whiskers now draw to the extreme points, not upper/lower fence
- Stopped using boxmean and instead used sdmultiple and sizemode 
- Implemented in a way so that diamond plots (boxmeas='sd') are impacted by sdmultiple

File: src/traces/box/calc.js [hunk 3]
```diff
@@ -286,7 +286,7 @@ module.exports = function calc(gd, trace) {
                 q1: _(gd, 'q1:'),
                 q3: _(gd, 'q3:'),
                 max: _(gd, 'max:'),
-                mean: trace.boxmean === 'sd' ? _(gd, 'mean ± σ:') : _(gd, 'mean:'),
+                mean: (trace.boxmean === 'sd') || (trace.sizemode === 'sd') ? (gd, 'mean ± ' + trace.sdmultiple + 'σ:') : _(gd, 'mean:'),
```
Comment (alexcjohnson) [inline src/traces/box/calc.js:289]: If `sdmultiple===1` let's omit it so the text still says `mean ± σ:`. And if we DO have a non-default multiple, can we insert it after the translation? That's the `_()`. Looking at the translations people have contributed, all of them keep `σ`, and with only one exception that I see they all keep `± σ`, so I'm thinking if we do something like:
```js
_(gd, 'mean ± σ:').replace(
    'σ',
    trace.sdmultiple===1 ? 'σ' : (trace.sdmultiple + 'σ')
)
```
(or pull it out to its own line so we don't need the noop replace in the default case) then these will all show the desired result.

The one exception is [zh-cn](https://github.com/plotly/plotly.js/blob/e8241996142b2612a19e09cdc00b29b896b853d1/lib/locales/zh-cn.js#L51) that uses `平均数 ± 标准差σ:` which would put the multiple in between `标准差` (seems to translate to standard deviation, ie they're clarifying what `σ` means) and `σ`, which would be a little weird but this is an edge case, we'd still be getting the default case (sdmultiple=1) correct. I suppose if we want to be REALLY thorough we could detect that there's something other than a space between `±` and `σ` and instead `.replace('σ', 'σ×' + trace.sdmultiple)` but at that point I'm kind of losing confidence that this (or any other option like trying to put the multiple after `± `) would be what Chinese readers want - any Chinese readers out there feel free to chime in!

File: src/traces/box/defaults.js [hunk 4]
```diff
@@ -43,6 +43,9 @@ function supplyDefaults(traceIn, traceOut, defaultColor, layout) {
     coerce('boxmean', boxmeanDflt);
 
     coerce('whiskerwidth');
+    coerce('showwhiskers');
+    coerce('sizemode');
```
Comment (alexcjohnson) [inline src/traces/box/defaults.js:47]:
```suggestion
    var sizemode = coerce('sizemode');
    coerce('showwhiskers', sizemode === 'quartiles');
```
if I'm understanding correctly that most users of `sd` mode would not want to see whiskers? And then if we do this, we'd also want to take [`dflt`](https://github.com/plotly/plotly.js/pull/6698/files#diff-591038d6258ff010e19e85399d7314ab4024713b2b5892ef8bcf443307145cf4R406) out of the `showwhiskers` attribute definition and describe this logic in its `description`.

File: src/traces/box/plot.js [hunk 5]
```diff
@@ -94,12 +95,15 @@ function plotBoxAndWhiskers(sel, axes, trace, t, isStatic) {
 
         var posm0 = posAxis.l2p(lcenter - bdPos0 * nw) + bPosPxOffset;
         var posm1 = posAxis.l2p(lcenter + bdPos1 * nw) + bPosPxOffset;
-        var q1 = valAxis.c2p(d.q1, true);
-        var q3 = valAxis.c2p(d.q3, true);
+        var q1 = trace.sizemode === 'sd' ? valAxis.c2p(d.mean - 1 * d.sd, true) :
+                                           valAxis.c2p(d.q1, true);
```
Comment (alexcjohnson) [inline src/traces/box/plot.js:99]: not a big deal but this (and a few below) can be simplified a little (is there a purpose to the `1*` or was that left over from the other version?):
```suggestion
        var sdmode = trace.sizemode === 'sd';
        var q1 = valAxis.c2p(sdmode ? d.mean - d.sd : d.q1, true);
```

File: test/image/mocks/box_sizemode_sd.json [hunk 6]
```diff
@@ -0,0 +1,156 @@
+{
```
Comment (alexcjohnson) [inline test/image/mocks/box_sizemode_sd.json:1]: Beautiful test image 😍

File: src/traces/box/defaults.js (hunk 4 above)
Comment (alexcjohnson) [reply src/traces/box/defaults.js:47]: Oh also with `sizemode='sd'` the `boxmean` attribute just doesn't make sense, right? We'd just be drawing a diamond that matches the box. So let's move the `coerce('boxmean', boxmeanDflt);` from above down here and put it inside an `if(sizemode === 'quartiles')`

File: src/traces/box/attributes.js [hunk 7]
```diff
@@ -378,6 +401,15 @@ module.exports = {
         ].join(' ')
     },
 
+    showwhiskers: {
+        valType: 'boolean',
+        dflt: true,
+        editType: 'calc',
+        description: [
+            'Determines whether or not whiskers are visible'
```
Comment (alexcjohnson) [inline src/traces/box/attributes.js:409]:
```suggestion
        editType: 'calc',
        description: [
            'Determines whether or not whiskers are visible.',
            'Defaults to true for `sizemode` *quartiles*, false for *sd*.'
```

File: src/traces/box/defaults.js [hunk 8]
```diff
@@ -43,6 +43,9 @@ function supplyDefaults(traceIn, traceOut, defaultColor, layout) {
     coerce('boxmean', boxmeanDflt);
 
     coerce('whiskerwidth');
+    var sizemode = coerce('sizemode');
```
Comment (alexcjohnson) [inline src/traces/box/defaults.js:46]:
```suggestion

    coerce('whiskerwidth');
    var sizemode = coerce('sizemode');
    if (sizemode === 'quartiles') {
        coerce('boxmean', boxmeanDflt);
    }
```

File: src/traces/box/plot.js [hunk 9]
```diff
@@ -94,12 +95,15 @@ function plotBoxAndWhiskers(sel, axes, trace, t, isStatic) {
...
+        var sdmode = trace.sizemode === 'sd';
+        var q1 = valAxis.c2p(sdmode ? d.mean - d.sd : d.q1, true);
+        var q3 = trace.sizemode === 'sd' ? valAxis.c2p(d.mean + 1 * d.sd, true) :
+                                           valAxis.c2p(d.q3, true);
         // make sure median isn't identical to either of the
         // quartiles, so we can see it
         var m = Lib.constrain(
-            valAxis.c2p(d.med, true),
+            trace.sizemode === 'sd' ? valAxis.c2p(d.mean, true) :
+                                      valAxis.c2p(d.med, true),
```
Comment (alexcjohnson) [inline src/traces/box/plot.js:106]:
```suggestion
            valAxis.c2p(sdmode ? d.mean : d.med, true),
```

File: src/traces/box/plot.js [hunk 10]
```diff
@@ -109,7 +113,7 @@ function plotBoxAndWhiskers(sel, axes, trace, t, isStatic) {
         // - box always has d.lf, but boxpoints can be anything
         // - violin has d.lf and should always use it (boxpoints is undefined)
         // - candlestick has only min/max
-        var useExtremes = (d.lf === undefined) || (trace.boxpoints === false);
+        var useExtremes = (d.lf === undefined) || (trace.boxpoints === false) || (trace.sizemode === 'sd');
```
Comment (alexcjohnson) [inline src/traces/box/plot.js:116]:
```suggestion
        var useExtremes = (d.lf === undefined) || (trace.boxpoints === false) || sdmode;
```

File: src/traces/violin/attributes.js [hunk 11]
```diff
@@ -155,6 +155,7 @@ module.exports = {
     hovertemplate: boxAttrs.hovertemplate,
 
     quartilemethod: boxAttrs.quartilemethod,
+    showwhiskers: boxAttrs.showwhiskers,
```
Comment (alexcjohnson) [inline src/traces/violin/attributes.js:158]: Not clear to me that we really want to add `showwhiskers` to violins, but if we do, they should be defined inside the `box` container below (and then whatever is needed for that setting to make it to the drawing routine) with `dflt: true` and their own description.

Anyway I'd be fine with omitting this for now, in which case the condition during draw can be `if (showwhiskers !== false)` instead of `if (showwhiskers)` so that `undefined` behaves like `true`.

File: src/traces/candlestick/attributes.js [hunk 12]
```diff
@@ -51,6 +51,9 @@ module.exports = {
     hovertext: OHLCattrs.hovertext,
 
     whiskerwidth: extendFlat({}, boxAttrs.whiskerwidth, { dflt: 0 }),
+    showwhiskers: boxAttrs.showwhiskers,
+    sizemode: boxAttrs.sizemode,
```
Comment (alexcjohnson) [inline src/traces/candlestick/attributes.js:55]: I don't think candlesticks make sense without whiskers, and certainly `sizemode` doesn't apply here. So we'll need to find another way to make them draw properly without creating these attributes.

**Comment by 28raining** [comment]
Hey @alexcjohnson. The baselines for Violin and Candlesticks were failing, because inside box/plot.js "trace.showwhiskers" was undefined.

Do you have any proposals to fix? I agree that it's not intended to add showwhiskers and sizemode attributes to those plots.

I was considering this:
`var showWhiskers = (typeof trace.showwhiskers === 'undefined') ? true : trace.showwhiskers;`

but seems not in the spirit of the coerce methodology.

**Comment by 28raining** [comment]
Hey @alexcjohnson. The baselines for Violin and Candlesticks were failing, because inside box/plot.js "trace.showwhiskers" was undefined.

I see your earlier comment and started using !== false. Thanks!

The remaining failing test is [webgl-jasmine], but seems unrelated to these changes

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -146,6 +146,12 @@ Three additional helpers exist that are refreshed every second:
 There is also a search bar in the top right of the dashboard. This fuzzy-searches
 image mocks based on their file name and trace type.
 
+#### Step 6b: Create a mock to test new features
+Create a new JSON inside
+`test\image\mocks\`
+which you'll then be able to search from the test dashboard
```
Comment (archmoj) [inline CONTRIBUTING.md:152]: In several cases adding new image tests are not necessary.
So this step should be optional and it could be part of `Step 6`.
You may revert changes in `CONTRIBUTING.md` and possibly submit a separate PR to suggest improvements as well.
Thank you!

File: src/traces/box/calc.js [hunk 14]
```diff
@@ -286,7 +286,9 @@ module.exports = function calc(gd, trace) {
                 q1: _(gd, 'q1:'),
                 q3: _(gd, 'q3:'),
                 max: _(gd, 'max:'),
-                mean: trace.boxmean === 'sd' ? _(gd, 'mean ± σ:') : _(gd, 'mean:'),
+                mean: (trace.boxmean === 'sd') || (trace.sizemode === 'sd') ?
+                    _(gd, 'mean ± σ:').replace('σ', trace.sdmultiple === 1 ? 'σ' : (trace.sdmultiple + 'σ')) : // displaying mean +- Nσ whilst supporting translations
```
Comment (archmoj) [inline src/traces/box/calc.js:290]: This line is getting too long.
How about adding a`getMean(gd, trace)` and move & revise this logic there?

File: src/traces/box/attributes.js [hunk 15]
```diff
@@ -218,7 +218,30 @@ module.exports = {
             'right (left) for vertical boxes and above (below) for horizontal boxes'
         ].join(' ')
     },
-
+    sdmultiple: {
+        valType: 'number',
+        min: 0,
```
[the remark this answers is not in the document]
Comment (alexcjohnson) [reply src/traces/box/attributes.js:223]: what if you want 0.5σ? Might be weird but I don't see a reason to prohibit it.

File: src/traces/box/calc.js (hunk 14 above)
Comment (28raining) [reply src/traces/box/calc.js:290]: Putting the code into a function would make the code less readable - if readability is the concern here.

File: src/traces/box/defaults.js [hunk 16]
```diff
@@ -40,9 +40,14 @@ function supplyDefaults(traceIn, traceOut, defaultColor, layout) {
...
     }
-    coerce('boxmean', boxmeanDflt);
 
     coerce('whiskerwidth');
+    var sizemode = coerce('sizemode');
+    if(sizemode === 'quartiles') {
+        coerce('boxmean', boxmeanDflt);
+    }
+    coerce('showwhiskers', sizemode === 'quartiles');
+    coerce('sdmultiple');
```
[the remark this answers is not in the document]
Comment (28raining) [reply src/traces/box/defaults.js:50]: Made this change. Guess you're saying performance of 2 if's >> coerce

**Comment by 28raining** [comment]
I've taken action on the comments

File: src/traces/box/defaults.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (alexcjohnson) [reply src/traces/box/defaults.js:50]: > Made this change

Pushed? I don't see it. Anyway it's not mainly about performance although performance will benefit. The goal is that if an attribute is not meaningful (based on other attributes) it should not be coerced at all so that it does not exist in `_fullData` (or `_fullLayout`). This helps with logic for drawing, as then `_fullData` is a precise description of what should be drawn, and it's also used by some external packages like [`react-chart-editor`](https://github.com/plotly/react-chart-editor) to decide what controls to display (if a control won't do anything it should not be shown).

File: src/traces/box/calc.js (hunk 14 above)
Comment (alexcjohnson) [reply src/traces/box/calc.js:290]: I agree - if it's only going to be used here, extracting to a function just means it takes another jump to read and understand the code. @archmoj if your concern is just line length we can break this onto more lines. Or leave it for now and sometime add `prettier` to our toolchain like we have in various other projects 😉

File: src/traces/box/defaults.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (28raining) [reply src/traces/box/defaults.js]: Sorry just pushed

**Review by alexcjohnson — APPROVED** [review]
💃 Looks good from my side - thanks @28raining!

**Comment by 28raining** [comment]
Wow, thanks everyone!
